### ISSUE TITLE
To be able to use the security layer of the database with connected users

### ISSUE THREAD

**Issue by Roman2nc** [metabase/metabase#25498]
**Is your feature request related to a problem? Please describe.**
Our organization's information system is already secure and several software programs use the database.
All our databases use the ldap authentication method (postgresql and clickhouse) and so does metabase.

we would like metabase to be able to use the authentication of the connected user during sql queries

**benefits:**

- the data sources are already secured by the database, no need to refine metabase for security

- the use of sql queries in metabase is already secured by the database, so there is no need to secure the query type questions in metabase

- other software can use the database with exactly the same security 

**disadvantages:**

- no query cache since the database carries the security

**Describe the solution you'd like**

Modification to be made on metabase:
- To be able to specify at the user account level its account linked to the datasource (if not filled in, then use the login)

- At the level of the security of the datasource in metabase, it is necessary to be able to configure the new mode of security by the database

- Use the user connected during the request to the database by applying a "set role".

**Configuration of administrator database:**

The database administrator of the organization is responsible for securing the data.

The administrator will apply directly at the database level the necessary scripts to secure the data.
He will create a "trust_user_roles" role in his database so that an administrator account can have exactly the same behavior as a user. 

At this stage, it does not concern metabase, just the system administrator

Creation of user accounts by the system administrator (script):

-- Add a specific role to be able to use the roles of a user
create role «trust_user_roles»;

create user «metabase_admin»;

create user «metabase_user_1»;
create user «metabase_user_2»;

grant «metabase_user_1» to «trust_user_roles»;
grant «metabase_user_2» to «trust_user_roles»;

-- We allow the metabase_admin account to pretend to be a user
grant «trust_user_roles» to «metabase_admin»;

When the metabase application will make an sql request on the datasource, it will be necessary to fix the role of the connected user in the same transaction.
 

//connected user
role_statement = «set role metabase_user_1»

query_statement = «select * from ...»

**Abnormal errors that can occur during the query**

- the database account does not have the right to use the role of the connected user (system administrator's configuration problem at the database level)

- the account of the connected user does not exist in the database (problem of configuration of the system administrator at the database level)

- there is no valid associated account for the connected user (problem in the configuration of metabase)

**Normal errors that can occur during the query**

- the connected user does not have the right to access an object of the request (access forbidden), that means that the user at the metabase level does not have the right to use this data

**Configuration of metabase:**

1 creation of the datasource

datasource «datawarehouse»
host: «localhost»
database: «drhouse»
user: «metabase_admin»
password: «****»

security: «database_security_layer»

2-creation of user accounts
- [user_1@myorg.nc](mailto:user_1@myorg.nc)
- [user_2@myorg.nc](mailto:user_1@myorg.nc)

3-association of accounts with data sources (optional, otherwise the login will be used)
[user_1@myorg.nc](mailto:user_1@myorg.nc) => [«datawarehouse»:metabase_user_1]
[user_2@myorg.nc](mailto:user_1@myorg.nc) => [«datawarehouse»:metabase_user_2]

**How important is this feature to you?**
This is very important in order not to short-circuit data security with two systems.
this feature will confirm our choice of the bi solution for our organization

**Comment by Roman2nc** [comment]
Yes, it's the same feature, I wanted to detail a solution that would work for us and tell you how important this functionality is.

Sorry for the duplication

**Comment by flamber** [comment]
Duplicate of #5147
